### PR TITLE
Upgrade to phpstan 0.12

### PR DESCRIPTION
**Pull request by spawnia**
I am particularly interested in the new generics feature that will allow us to resolve https://github.com/webonyx/graphql-php/issues/285

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -15,10 +15,10 @@
...
   "require-dev": {
     "doctrine/coding-standard": "^6.0",
-    "phpbench/phpbench": "^0.14.0",
-    "phpstan/phpstan": "0.11.16",
-    "phpstan/phpstan-phpunit": "0.11.2",
-    "phpstan/phpstan-strict-rules": "0.11.1",
+    "phpbench/phpbench": "^0.14",
+    "phpstan/phpstan": "^0.12",
+    "phpstan/phpstan-phpunit": "^0.12",
+    "phpstan/phpstan-strict-rules": "^0.12",
```
Comment (simPod) [inline composer.json]: We need to lock patch versions as well. New "patch" release might break build otherwise.

File: src/Language/AST/NodeList.php [hunk 2]
```diff
@@ -14,13 +14,16 @@
 use function count;
 use function is_array;
 
+/**
+ * @template T
```
Comment (simPod) [inline src/Language/AST/NodeList.php]: I'd use vendor prefix for these `@phpstan-template`

File: src/Language/AST/NodeList.php [hunk 3]
```diff
@@ -14,13 +14,16 @@
...
+ */
 class NodeList implements ArrayAccess, IteratorAggregate, Countable
 {
-    /** @var Node[]|mixed[] */
+    /** @var T[] */
```
Comment (simPod) [inline src/Language/AST/NodeList.php]: Same here, let's keep `@var` for IDE support and use `@phpstan-var` for generic type

File: src/Language/AST/NodeList.php (hunk 2 above)
Comment (simPod) [reply src/Language/AST/NodeList.php]: Shouldn't we also have

```php
@phpstan-implements ArrayAccess<T>
@phpstan-implements IteratorAggregate <T>
```

File: src/Language/AST/NodeList.php [hunk 4]
```diff
@@ -50,7 +53,7 @@ public function offsetExists($offset)
     /**
      * @param mixed $offset
      *
-     * @return mixed
+     * @return T
```
Comment (simPod) [inline src/Language/AST/NodeList.php]: If we add `implements` annotations, this might be possible to drop.

File: composer.json (hunk 1 above)
Comment (simPod) [reply composer.json]: `0.12.0`

File: src/Language/AST/NodeList.php (hunk 3 above)
Comment (spawnia) [reply src/Language/AST/NodeList.php]: So duplicate the types?

**Comment by spawnia** [comment]
@simPod i am stuck with the remaining issues. I invited you as a collaborator on my fork, maybe you can give it a shot.

File: phpstan.neon.dist [hunk 5]
```diff
@@ -25,6 +25,11 @@ parameters:
         - "~Access to an undefined property GraphQL\\\\Language\\\\AST\\\\.+::\\$didEnter~"
         - "~Access to an undefined property GraphQL\\\\Language\\\\AST\\\\.+::\\$didLeave~"
         - "~Access to an undefined property GraphQL\\\\Language\\\\AST\\\\Node::\\$value~"
+        # TODO fix those newly added errors and finish upgrading to PhpStan 0.12
```
Comment (spawnia) [inline phpstan.neon.dist]: I think we can do those in separate PR's, since those throw over a hundred errors if you combine them.

**Comment by simPod** [comment]
@spawnia will take a look tmrw

File: src/Error/Error.php [hunk 6]
```diff
@@ -171,7 +171,7 @@ public static function createLocatedError($error, $nodes = null, $path = null)
             $message = (string) $error;
         }
 
-        return new static(
+        return new self(
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Error/Error.php]: Right. I must say i don't really like the phpstan rule behind this for library code. Locking down classes in very rigid ways hurts reusability and flexibility.

**Comment by spawnia** [comment]
@simPod great job at getting the `NodeList` working. You have a really good understanding of PhpStan!

I did clean up the phpstan config a bit, we can iterate on the smaller issues piece by piece from now on.

**Comment by shmax** [comment]
It's nice to see a collaborative branch like this take shape. 👍 

Are either of you interested in teaming up on PHPStan level 4?

**Comment by spawnia** [comment]
> It's nice to see a collaborative branch like this take shape. 
> 
> Are either of you interested in teaming up with me on PHPStan level 4?

Definitely interested in leveling up further. There are a whole bunch of `ignoreErrors` that i think we should try and fix before piling on even more from level 4.

**Comment by shmax** [comment]
> There are a whole bunch of ignoreErrors that i think we should try and fix before piling on even more from level 4.

Sounds good.

File: src/Language/AST/NodeList.php [hunk 7]
```diff
@@ -6,101 +6,127 @@
...
      */
-    public function offsetExists($offset)
+    public function offsetExists($offset) : bool
     {
         return isset($this->nodes[$offset]);
     }
 
     /**
-     * @param mixed $offset
+     * TODO this *should* always return a Node, but sometimes a string is in here?
```
Comment (spawnia) [inline src/Language/AST/NodeList.php]: @vladar can you enlighten us on what is going on here?

During some tests, the `Printer` puts items such as `"input:SomeInput"` in a NodeList and mutates it somehow? That makes static validation quite hard/impossible, since we can no longer guarantee that the type `T` is always the same.

**Comment by spawnia** [comment]
> phpstan v0.12.1 just came out few minutes ago, can bump ;)

Done, they even fixed a bug.

File: src/Language/AST/NodeList.php (hunk 7 above)
Comment (vladar) [reply src/Language/AST/NodeList.php]: Printer uses a Visitor which can transform any node to an arbitrary data structure (in case of printer it transforms a node to a printed version). See here:

https://github.com/webonyx/graphql-php/blob/34fe9615097be1f1d7f628ef563489973905e759/src/Language/Visitor.php#L31-L58

Technically visitor doesn't mutate but creates a new structure (clones where appropriate). I guess for strict typing we will have to adjust Visitor and replace NodeList with array (or less strict List structure when modifications are made).

Comment (spawnia) [reply src/Language/AST/NodeList.php]: Switching to an array when transforming seems fine, since i think the lazyness aspect of `NodeList` does not matter at this point anymore.

How about we open a separate PR for that and merge this one as is to enable further work on phpstan:0.12 compatibility?

Comment (vladar) [reply src/Language/AST/NodeList.php]: Makes sense, we have a TODO in place there so can revisit when Visitor part is unblocked

File: src/Error/Error.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Error/Error.php]: @vladar i think the route we choose to go on these issues is an important design decision. There are basically two paths we can go down.

We can lock everything down to the max, making all classes final and methods preferrably private. This does prevent misuse and keeps the API surface small, but hurts flexibility and extensibility. I personally think this approach makes more sense in large project codebases where you want to keep core components clean and limit technical debt.

The other way would be to keep classes open, allowing extending them and overwriting or calling protected methods. The BC guarantees are only upheld for the parts of the library marked with `@api`. While this provides some footguns and may cause pain for developers who choose to mess with the insides of this library, i think that this is not really a drawback from a library perspective. It gives more power and responsibility to users of this library.

[the remark this answers is not in the document]
Comment (vladar) [reply src/Error/Error.php]: I agree with you here. I think it is quite an impractical rule. `@api` convention should be enough to communicate which APIs we consider `stable`. Then the risk of overriding something is on developers messing with internals.

I know this feeling when library authors didn't consider a specific use-case and locked-down everything so that you couldn't even "fix" it on the user-land without forking or some crazy reflection hacks.

[the remark this answers is not in the document]
Comment (spawnia) [reply src/Error/Error.php]: I reverted the breaking changes here for now, we can continue this discussion in https://github.com/webonyx/graphql-php/issues/593

File: src/Validator/Rules/KnownDirectives.php [hunk 8]
```diff
@@ -186,6 +188,8 @@ private function getDirectiveLocationForASTPath(array $ancestors)
                     ? DirectiveLocation::INPUT_FIELD_DEFINITION
                     : DirectiveLocation::ARGUMENT_DEFINITION;
         }
+
+        throw new Exception('Unknown directive location: ' . get_class($appliedTo));
```
Comment (simPod) [inline src/Validator/Rules/KnownDirectives.php]: Maybe something like `InvariantViolation::shouldNotHappen()` can be used?
